File: src/test/java/io/lsdconsulting/lsd/distributed/mongo/integration/InterceptedDocumentMongoRepositoryIT.java
package io.lsdconsulting.lsd.distributed.mongo.integration;

import com.mongodb.client.ListIndexesIterable;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.TestApplication;
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.config.RepositoryConfig;
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.repository.TestRepository;
import io.lsdconsulting.lsd.distributed.mongo.repository.InterceptedDocumentMongoRepository;
import org.bson.Document;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;

import java.time.ZoneId;
import java.util.List;
import java.util.stream.Collectors;

import static io.lsdconsulting.lsd.distributed.access.model.Type.REQUEST;
import static io.lsdconsulting.lsd.distributed.mongo.integration.testapp.repository.TestRepository.*;
import static java.time.Instant.ofEpochSecond;
import static java.time.ZonedDateTime.ofInstant;
import static java.util.stream.StreamSupport.stream;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;

@Import(RepositoryConfig.class)
@SpringBootTest(webEnvironment = RANDOM_PORT, classes = {TestApplication.class})
class InterceptedDocumentMongoRepositoryIT {

    @Autowired
    private TestRepository testRepository;

    private InterceptedDocumentMongoRepository underTest;

    @BeforeEach
    void setup() {
        underTest = new InterceptedDocumentMongoRepository("mongodb://" + MONGODB_HOST + ":" + MONGODB_PORT,  1500, 1L);
    }

    @AfterAll
    static void tearDown() {
        tearDownDatabase();
        tearDownClient();
    }

    @Test
    public void shouldCreateCollectionWithIndexes() {
        final ListIndexesIterable<Document> indexes = testRepository.getCollection().listIndexes();
        final List<String> indexNames = stream(indexes.spliterator(), false)
                .map(doc -> (String) doc.get("name"))
                .collect(Collectors.toList());

        assertThat(indexNames, hasItem(startsWith("traceId")));
        assertThat(indexNames, hasItem(startsWith("createdAt")));
    }

    @Test
    public void shouldCreateCappedCollection() {
        final boolean isCapped = testRepository.isCollectionCapped();

        assertTrue(isCapped);
    }

    @Test
    public void shouldSaveAndRetrieveFromCollection() {
        InterceptedInteraction interceptedInteraction = InterceptedInteraction.builder()
                .elapsedTime(20L)
                .httpStatus("OK")
                .path("/path")
                .httpMethod("GET")
                .body("body")
                .type(REQUEST)
                .traceId("traceId")
                .createdAt(ofInstant(ofEpochSecond(0), ZoneId.of("UTC")))
                .build();

        underTest.save(interceptedInteraction);

        List<InterceptedInteraction> result = underTest.findByTraceIds("traceId");

        assertThat(result, hasSize(1));
        assertThat(result.get(0).getElapsedTime(), is(20L));
        assertThat(result.get(0).getHttpStatus(), is("OK"));
        assertThat(result.get(0).getPath(), is("/path"));
        assertThat(result.get(0).getHttpMethod(), is("GET"));
        assertThat(result.get(0).getBody(), is("body"));
        assertThat(result.get(0).getType(), is(REQUEST));
        assertThat(result.get(0).getCreatedAt(), is(ofInstant(ofEpochSecond(0), ZoneId.of("UTC"))));
    }
}

File: src/test/java/io/lsdconsulting/lsd/distributed/mongo/integration/InterceptedDocumentMongoRepositoryIT.kt
package io.lsdconsulting.lsd.distributed.mongo.integration

import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.TestApplication
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.config.RepositoryConfig
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.repository.TestRepository
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.repository.TestRepository.Companion.tearDownClient
import io.lsdconsulting.lsd.distributed.mongo.integration.testapp.repository.TestRepository.Companion.tearDownDatabase
import io.lsdconsulting.lsd.distributed.mongo.repository.InterceptedDocumentMongoRepository
import org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric
import org.bson.Document
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.jeasy.random.EasyRandom
import org.jeasy.random.EasyRandomParameters
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.context.annotation.Import
import java.time.Instant.ofEpochSecond
import java.time.ZoneId
import java.time.ZonedDateTime.now
import java.time.ZonedDateTime.ofInstant
import java.time.temporal.ChronoUnit
import java.util.stream.Collectors
import java.util.stream.StreamSupport


@Import(RepositoryConfig::class)
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT, classes = [TestApplication::class])
internal class InterceptedDocumentMongoRepositoryIT {
    private val easyRandom = EasyRandom(EasyRandomParameters().seed(System.currentTimeMillis()))

    @Autowired
    private lateinit var testRepository: TestRepository

    private lateinit var underTest: InterceptedDocumentMongoRepository

    @BeforeEach
    fun setup() {
        underTest = InterceptedDocumentMongoRepository(
            "mongodb://" + TestRepository.MONGODB_HOST + ":" + TestRepository.MONGODB_PORT,
            1500,
            1L
        )
    }

    @Test
    fun `should create collection with indexes`() {
        val indexes = testRepository.collection.listIndexes()
        val indexNames = StreamSupport.stream(indexes.spliterator(), false)
            .map { doc: Document -> doc["name"] as String? }
            .collect(Collectors.toList())
        assertThat(indexNames, hasItem(startsWith("traceId")))
        assertThat(indexNames, hasItem(startsWith("createdAt")))
    }

    @Test
    fun `should create capped collection`() {
        val isCapped = testRepository.isCollectionCapped
        Assertions.assertTrue(isCapped)
    }

    @Test
    fun `should save and retrieve from collection`() {
        val interceptedInteraction = InterceptedInteraction(
            elapsedTime = 20L,
            httpStatus = "OK",
            path = "/path",
            httpMethod = "GET",
            body = "body",
            interactionType = InteractionType.REQUEST,
            traceId = "traceId",
            createdAt = ofInstant(ofEpochSecond(0), ZoneId.of("UTC")))

        underTest.save(interceptedInteraction)

        val result = underTest.findByTraceIds("traceId")
        assertThat(result, hasSize(1))
        assertThat(result[0].elapsedTime, `is`(20L))
        assertThat(result[0].httpStatus, `is`("OK"))
        assertThat(result[0].path, `is`("/path"))
        assertThat(result[0].httpMethod, `is`("GET"))
        assertThat(result[0].body, `is`("body"))
        assertThat(result[0].interactionType, `is`(InteractionType.REQUEST))
        assertThat(result[0].createdAt, `is`(ofInstant(ofEpochSecond(0), ZoneId.of("UTC"))))
    }

    @Test
    fun `should save and retrieve random entry from collection`() {
        val traceId = randomAlphanumeric(10)
        val interceptedInteraction = buildInterceptedInteraction(traceId)

        underTest.save(interceptedInteraction)

        val result = underTest.findByTraceIds(interceptedInteraction.traceId)
        assertThat(result, hasSize(1))
        assertThat(result[0].elapsedTime, `is`(interceptedInteraction.elapsedTime))
        assertThat(result[0].httpStatus, `is`(interceptedInteraction.httpStatus))
        assertThat(result[0].path, `is`(interceptedInteraction.path))
        assertThat(result[0].httpMethod, `is`(interceptedInteraction.httpMethod))
        assertThat(result[0].body, `is`(interceptedInteraction.body))
        assertThat(result[0].interactionType, `is`(interceptedInteraction.interactionType))
        assertThat(result[0].createdAt, `is`(interceptedInteraction.createdAt))
    }

    @Test
    fun `should save and retrieve in correct order`() {
        val traceId = randomAlphanumeric(10)
        val interceptedInteractions = (1..10)
            .map { _ -> buildInterceptedInteraction(traceId) }
            .sortedByDescending { it.createdAt }

        interceptedInteractions.forEach { underTest.save(it) }

        val result = underTest.findByTraceIds(traceId)
        assertThat(result, hasSize(10))
        (1..10).forEach { assertThat(result[it -1].createdAt, `is`(interceptedInteractions[10 - it].createdAt)) }
    }

    private fun buildInterceptedInteraction(traceId: String) = easyRandom.nextObject(InterceptedInteraction::class.java)
        .copy(traceId = traceId, createdAt = now(ZoneId.of("UTC")).truncatedTo(ChronoUnit.MILLIS))

    companion object {
        @JvmStatic
        @AfterAll
        fun tearDown() {
            tearDownDatabase()
            tearDownClient()
        }
    }
}
